File: libs/common/estategap_common/models/scoring.py
"""ScoringResult and SHAP attribution models."""

from __future__ import annotations

from decimal import Decimal
from enum import IntEnum
from uuid import UUID

from pydantic import field_validator

from ._base import AwareDatetime, EstateGapModel, validate_country_code


class DealTier(IntEnum):
    """Discrete deal quality buckets emitted by the scorer."""

    GREAT_DEAL = 1
    GOOD_DEAL = 2
    FAIR = 3
    OVERPRICED = 4


class ShapValue(EstateGapModel):
    """Top-N SHAP attribution value for a scored listing."""

    feature_name: str
    value: float


class ScoringResult(EstateGapModel):
    """ML scoring payload merged into a persisted listing."""

    listing_id: UUID
    country: str
    estimated_price: Decimal
    deal_score: Decimal
    deal_tier: DealTier
    confidence_low: Decimal
    confidence_high: Decimal
    shap_features: list[ShapValue]
    model_version: str
    scored_at: AwareDatetime

    @field_validator("country")
    @classmethod
    def _validate_country(cls, value: str) -> str:
        return validate_country_code(value)


__all__ = ["DealTier", "ScoringResult", "ShapValue"]

File: libs/pkg/models/scoring.go
package models

import (
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/shopspring/decimal"
)

type ShapValue struct {
	FeatureName string  `json:"feature_name" db:"feature_name"`
	Value       float64 `json:"value" db:"value"`
}

type ScoringResult struct {
	ListingID      pgtype.UUID        `json:"listing_id" db:"listing_id"`
	Country        string             `json:"country" db:"country"`
	EstimatedPrice decimal.Decimal    `json:"estimated_price" db:"estimated_price"`
	DealScore      decimal.Decimal    `json:"deal_score" db:"deal_score"`
	DealTier       DealTier           `json:"deal_tier" db:"deal_tier"`
	ConfidenceLow  decimal.Decimal    `json:"confidence_low" db:"confidence_low"`
	ConfidenceHigh decimal.Decimal    `json:"confidence_high" db:"confidence_high"`
	ShapFeatures   []ShapValue        `json:"shap_features" db:"shap_features"`
	ModelVersion   string             `json:"model_version" db:"model_version"`
	ScoredAt       pgtype.Timestamptz `json:"scored_at" db:"scored_at"`
}
